### PR TITLE
client/finality-grandpa: Make round_communication use bounded channel

### PR DESCRIPTION
**Pull request by mxinden**
`round_communication` returns a `Sink` and a `Stream` for outgoing and
incoming messages. The messages send into the `Sink` are forwarded down
to the network as well as send back into the `Stream` to ensure the node
processes its own messages.

So far, to send messages into the `Sink` back into the `Stream`, an
unbounded channel was used. This patch updates `round_communication` and
`OutgoingMessages` to use a bounded channel.

This is part of a greater effort to reduce the number of owners of
components within `finality-grandpa` and `network` as well as to reduce
the amount of unbounded channels. For details see d4fbb897c and
f0c18520a.

### CODE REVIEW

**Comment by mxinden** [comment]
> We need to ensure that we don’t create deadlocks.

Correct. The channel being bounded in this patch is used to send packets back to our self. Thus we depend on us reading from the channel.

**Comment by tomaka** [comment]
Let's maybe merge https://github.com/paritytech/substrate/pull/4612 first?

**Comment by DemiMarie-parity** [comment]
@mxinden We need to ensure that we don’t buffer an unbounded amount of data.

**Comment by andresilva** [comment]
I agree with the theory behind this change. In practice though I think this is bound to cause more problems than it aims to solve. In particular because AFAIK going OOM due to these unbounded channels has not been a problem so far. I'd appreciate that this is properly tested before it is rolled out on a live network to avoid surprises (maybe by deploying to flaming fir along with #4612).

**Comment by mxinden** [comment]
> @mxinden We need to ensure that we don’t need to buffer an unbounded amount of data.

@DemiMarie-parity I don't understand your comment. Yes, we need to make sure to not buffer unbounded amount of data. That is what the effort behind these pull requests tries to achieve. In the end you can see an unbounded channel as an unbounded buffer.

**Comment by mxinden** [comment]
I would prefer not to block this pull request on https://github.com/paritytech/substrate/pull/4612 given that the latter is quite large and given that it is unclear when it will be merged.

> so that we don't need to mix futures versions.

We already mix both futures01 and futures03 within `client/finality-grandpa`.

The changeset in this pull request is not large and I am happy to resolve the merge conflicts with https://github.com/paritytech/substrate/pull/4612 (already half done locally) once this one merged.

@tomaka @andresilva @expenses let me know if this would be ok for you.

(I will look into https://github.com/paritytech/substrate/pull/4612 and https://github.com/paritytech/substrate/pull/4633 next and try to help out.)

File: client/finality-grandpa/src/communication/mod.rs [hunk 1]
```diff
@@ -290,22 +291,20 @@ impl<B: BlockT, N: Network<B>> NetworkBridge<B, N> {
...
-				decoded.ok()
-			})
-			.and_then(move |msg| {
-				match msg {
-					GossipMessage::Vote(msg) => {
+
+				match decoded {
+					Err(ref e) => {
+						debug!(target: "afg", "Skipping malformed message {:?}: {}", notification, e);
+						return ready(None);
```
Comment (andresilva) [inline client/finality-grandpa/src/communication/mod.rs]: I'd prefer if this read `future::ready` instead. Wasn't immediately obvious to me.

**Comment by andresilva** [comment]
> We already mix both futures01 and futures03 within client/finality-grandpa.

After #4612 we don't, which is what I meant. I'm OK with not blocking this PR though.

File: client/finality-grandpa/src/communication/mod.rs (hunk 1 above)
Comment (rphmeier) [reply client/finality-grandpa/src/communication/mod.rs]: I agree. As a rule of thumb, importing free functions into the local namespace hurts readability.

`module::function` lets us know where to look for the function without searching the imports.
`function` lets us know that the function is in the current module.

Comment (mxinden) [reply client/finality-grandpa/src/communication/mod.rs]: Good point. Fixed with https://github.com/paritytech/substrate/pull/4691/commits/6858319a169babd168871414883bbb969cc8c32d.

**Comment by mxinden** [comment]
Thanks for the reviews and help!

@rphmeier any further comments from your side?

**Pull request merged**